### PR TITLE
Use yotta config for declaring driver support in `device.h`.

### PR DESCRIPTION
**Pull request by niklas-arm**
This is the umbrella PR for moving the `device.h` and `target_config.h` out of the chip specific hals into the target configuration using yotta's config system.
This PR needs to be synchronised with the chip hals and targets.

This is WIP. Related PRs:

1. Update yt config in all targets:
    - [ ] STM32F429-DISCO: https://github.com/ARMmbed/target-st-stm32f429i-disco/pull/26
    - [ ] STM32F401-NUCLEO:
    - [ ] Freescale K64F:
    - [ ] Nordic NRF51:
1. Move defines with defaults into repos:
    - [ ] move `device.h` into `mbed-hal`: This PR
    - [ ] move `target_config.h/MODULE_SIZE_*` into `mbed-hal`: This PR
    - [ ] move `target_config.h/SPI_TRANSACTION_QUEUE_SIZE_*` into `mbed-drivers`:
    - [ ] move `target_config.h/MINAR_PLATFORM_*` into `minar-platform`:
1. Remove `device.h` and `target_config.h` from chip specific hal:
    - [ ] STM32F4:
    - [ ] K64F:
    - [ ] NRF51:

@0xc0170 @bremoran @autopulated @bogdanm

### CODE REVIEW

File: mbed-hal/device.h [hunk 1]
```diff
@@ -0,0 +1,159 @@
+/* mbed Microcontroller Library
+ *******************************************************************************
+ * Copyright (c) 2014, STMicroelectronics
```
Comment (0xc0170) [inline mbed-hal/device.h]: change the license text :-)

**Comment by bremoran** [comment]
In general, +1

Can you outline the strategy to make this a non-breaking change?

**Comment by niklas-arm** [comment]
The strategy is:
1. provide the correct configuration settings in the targets and publish them. This has no effect.
1. move the defines into the respective repos (`mbed-hal`, `mbed-drivers` & `minar-platform`) and initialize them with useful defaults like these changes. `mbed-hal/device.h` should now contain the same settings as `mbed-hal-chip/device.h`, so regardless which file is chosen, it should result in the same. Care needs to be taken for `target_config.h`: the defines in there need to be guarded in their new place.
1. Remove `device.h` and `target_config.h` from `mbed-hal-chip`, their function should now seamlessly by taken over by yt config.
1. Profit

**Comment by autopulated** [comment]
Process looks good. Could you add a checklist here of the steps, including which targets need updating?

**Comment by niklas-arm** [comment]
The meaning of this PR has changed. See #69.

**Pull request closed without merging**